`-Xcheck-macros` regression in `upickle` dependent projects

Based on the OpenCB failures in: 
- automorph-org/automorph - [build logs](https://scala3.westeurope.cloudapp.azure.com/dashboard/projects/automorph-org/automorph/builds/scheduled_3.8.4-RC1-bin-20260313-3be6338-NIGHTLY/logs)
- kitlangton/neotype - [build logs](https://scala3.westeurope.cloudapp.azure.com/dashboard/projects/kitlangton/neotype/builds/scheduled_3.8.4-RC1-bin-20260313-3be6338-NIGHTLY/logs) 

## Compiler version

Last good release: 3.8.4-RC1-bin-20260225-516f4d7-NIGHTLY
First bad release: 3.8.4-RC1-bin-20260226-c0af6f4-NIGHTLY
Bisect points to 46ca4912141633e7145fadb6ff3fc8bc9ec4cec0  / https://github.com/scala/scala3/pull/24922

## Minimized code

```scala
//> using options -Xcheck-macros
//> using file test-macro.scala

final case class MessageError[Node](message: Option[String], code: Option[Int], data: Option[Node])

object Test:
  def readWriter(custom: Custom): Unit =
    custom.macroRW[MessageError[Int]]
```

```scala
// test-macro.scala 

import scala.deriving.Mirror
import scala.quoted.*

object LocalMacros:
  def macroRWImpl[T: Type](using Quotes): Expr[Unit] =
    import quotes.reflect.*

    def inspect(tpe: TypeRepr, typeArgs: List[TypeRepr]): Unit =
      val constructorSym = tpe.typeSymbol.primaryConstructor
      val constructorParamSymss = constructorSym.paramSymss
      val (tparams0, params0) = constructorParamSymss.flatten.partition(_.isType)
      val constructorTpe = tpe.memberType(constructorSym).widen
      params0.foreach { param =>
        constructorTpe.memberType(param).substituteTypes(tparams0, typeArgs)
      }

    TypeRepr.of[T] match
      case applied: AppliedType => inspect(TypeRepr.of[T], applied.args)
      case _ => inspect(TypeRepr.of[T], Nil)

    '{ () }

trait Custom:
  inline def macroRW[T](using Mirror.Of[T]): Unit = ${ LocalMacros.macroRWImpl[T] }

```

## Output

```scala
[error] ./test.scala:8:5
[error] Exception occurred while executing macro expansion.
[error] java.lang.AssertionError: val message is not a member of [Node]
[error]   (message: Option[String], code: Option[Int], data: Option[Node]):
[error]     MessageError[Node]
[error]         at LocalMacros$.inspect$1$$anonfun$1(test-macro.scala:14)
[error]         at scala.collection.immutable.List.foreach(List.scala:327)
[error]         at LocalMacros$.inspect$1(test-macro.scala:13)
[error]         at LocalMacros$.macroRWImpl(test-macro.scala:18)
[error] 
[error]     custom.macroRW[MessageError[Int]]
[error]     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Error compiling project (Scala 3.8.4-RC1-bin-20260317-9a56448-NIGHTLY, JVM (21))
```

## Expectation
Should compile without assertion error?

--- a/compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala
+++ b/compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala
@@ -1926,7 +1926,7 @@ class QuotesImpl private (using val ctx: Context) extends Quotes, QuoteUnpickler
               boundary.break(false)
             })
           }
-          xCheckMacroWarning(isTypeRelatedToThePassedMember, s"$member is not a valid member of ${self.show}. Returned type might not be the one expected.")
+          xCheckMacroAssert(isTypeRelatedToThePassedMember, s"$member is not a valid member of ${self.show}")
 
           // we replace thisTypes here to avoid resolving otherwise unstable prefixes into Nothing
           val memberInfo =
@@ -3550,17 +3550,6 @@ class QuotesImpl private (using val ctx: Context) extends Quotes, QuoteUnpickler
     private inline def xCheckMacroAssert(inline cond: Boolean, inline msg: String): Unit =
       if xCheckMacro && !cond then
         xCheckMacroAssertFail(msg)
-
-    private inline def xCheckMacroWarning(inline cond: Boolean, inline msg: String): Unit =
-      if xCheckMacro && !cond then
-        val error = new AssertionError()
-        if !yDebugMacro then
-          // start stack trace at the place where the user called the reflection method
-          error.setStackTrace(
-            error.getStackTrace
-              .dropWhile(_.getClassName().startsWith("scala.quoted.runtime.impl")))
-        val stackTrace = error.getStackTrace().map(_.toString).mkString("\n")
-        report.warning(s"$msg\n${stackTrace}")
 
     private def xCheckMacroAssertFail(msg: String): Unit =
       val error = new AssertionError(msg)

--- a/compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala
+++ b/compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala
@@ -28,7 +28,6 @@ import dotty.tools.dotc.core.NameKinds.ExceptionBinderName
 import dotty.tools.dotc.transform.TreeChecker
 import dotty.tools.dotc.core.Names
 import dotty.tools.dotc.util.Spans.NoCoord
-import dotty.tools.dotc.report
 
 object QuotesImpl {
 
@@ -1905,17 +1904,14 @@ class QuotesImpl private (using val ctx: Context) extends Quotes, QuoteUnpickler
           //     (and failed dotty compilation tests)
           // Additionally, it can be useful to be able to learn a type of a symbol using some indirect prefix,
           // even if that symbol is not a direct member of that prefix, but a nested one.
-          def isTypeRelatedToThePassedMember = {
+          def isTypeRelatedToThePassedMember =
             import scala.util.boundary
-
-            // the symbol will not exist for LambdaTypes,
+            // the symbol for `self` will not exist for LambdaTypes,
             // which don't seem to be supported by asSeenFrom anyway
-            // (but don't seem to cause crashes, and return no-op member.info,
-            // which we might not have access other ways).
-            val selfSymExists = self.typeSymbol.exists
+            // (but also don't seem to cause crashes, and return no-op member.info,
+            // which we might not be able to access otherwise, so we need to allow this).
             val isLambdaType = self.isInstanceOf[LambdaType]
-            (!selfSymExists && isLambdaType) ||
-            (selfSymExists && boundary {
+            isLambdaType || boundary {
               var checked: Symbol = member
               while(checked.exists) {
                 if self.derivesFrom(checked)
@@ -1925,9 +1921,9 @@ class QuotesImpl private (using val ctx: Context) extends Quotes, QuoteUnpickler
                 checked = checked.owner
               }
               boundary.break(false)
-            })
-          }
-          xCheckMacroAssert(isTypeRelatedToThePassedMember, s"$member is not a valid member of ${self.show}")
+            }
+
+          xCheckMacroAssert(isTypeRelatedToThePassedMember, s"$member is not a member of ${self.show}")
 
           // we replace thisTypes here to avoid resolving otherwise unstable prefixes into Nothing
           val memberInfo =

--- a/compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala
+++ b/compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala
@@ -1906,13 +1906,14 @@ class QuotesImpl private (using val ctx: Context) extends Quotes, QuoteUnpickler
           // Additionally, it can be useful to be able to learn a type of a symbol using some indirect prefix,
           // even if that symbol is not a direct member of that prefix, but a nested one.
           def isTypeRelatedToThePassedMember = {
+            import scala.util.boundary
+
             // the symbol will not exist for LambdaTypes,
             // which don't seem to be supported by asSeenFrom anyway
             // (but don't seem to cause crashes, and return no-op member.info,
             // which we might not have access other ways).
             val selfSymExists = self.typeSymbol.exists
             val isLambdaType = self.isInstanceOf[LambdaType]
-            import scala.util.boundary
             (!selfSymExists && isLambdaType) ||
             (selfSymExists && boundary {
               var checked: Symbol = member
@@ -1934,7 +1935,7 @@ class QuotesImpl private (using val ctx: Context) extends Quotes, QuoteUnpickler
               member.info.substThis(self.classSymbol.asClass, self)
             else
               member.info
-          println("memInfo " + member.info)
+
           memberInfo.asSeenFrom(self, member.owner) match
             case dotc.core.Types.ClassInfo(prefix, sym, _, _, _) =>
               // We do not want to expose ClassInfo in the reflect API, instead we change it to a TypeRef,

--- a/compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala
+++ b/compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala
@@ -28,6 +28,7 @@ import dotty.tools.dotc.core.NameKinds.ExceptionBinderName
 import dotty.tools.dotc.transform.TreeChecker
 import dotty.tools.dotc.core.Names
 import dotty.tools.dotc.util.Spans.NoCoord
+import dotty.tools.dotc.report
 
 object QuotesImpl {
 
@@ -1904,9 +1905,16 @@ class QuotesImpl private (using val ctx: Context) extends Quotes, QuoteUnpickler
           //     (and failed dotty compilation tests)
           // Additionally, it can be useful to be able to learn a type of a symbol using some indirect prefix,
           // even if that symbol is not a direct member of that prefix, but a nested one.
-          def isTypeRelatedToThePassedMember =
+          def isTypeRelatedToThePassedMember = {
+            // the symbol will not exist for LambdaTypes,
+            // which don't seem to be supported by asSeenFrom anyway
+            // (but don't seem to cause crashes, and return no-op member.info,
+            // which we might not have access other ways).
+            val selfSymExists = self.typeSymbol.exists
+            val isLambdaType = self.isInstanceOf[LambdaType]
             import scala.util.boundary
-            boundary {
+            (!selfSymExists && isLambdaType) ||
+            (selfSymExists && boundary {
               var checked: Symbol = member
               while(checked.exists) {
                 if self.derivesFrom(checked)
@@ -1916,8 +1924,9 @@ class QuotesImpl private (using val ctx: Context) extends Quotes, QuoteUnpickler
                 checked = checked.owner
               }
               boundary.break(false)
-            }
-          xCheckMacroAssert(isTypeRelatedToThePassedMember, s"$member is not a member of ${self.show}")
+            })
+          }
+          xCheckMacroWarning(isTypeRelatedToThePassedMember, s"$member is not a valid member of ${self.show}. Returned type might not be the one expected.")
 
           // we replace thisTypes here to avoid resolving otherwise unstable prefixes into Nothing
           val memberInfo =
@@ -1925,7 +1934,7 @@ class QuotesImpl private (using val ctx: Context) extends Quotes, QuoteUnpickler
               member.info.substThis(self.classSymbol.asClass, self)
             else
               member.info
-
+          println("memInfo " + member.info)
           memberInfo.asSeenFrom(self, member.owner) match
             case dotc.core.Types.ClassInfo(prefix, sym, _, _, _) =>
               // We do not want to expose ClassInfo in the reflect API, instead we change it to a TypeRef,
@@ -3541,6 +3550,17 @@ class QuotesImpl private (using val ctx: Context) extends Quotes, QuoteUnpickler
     private inline def xCheckMacroAssert(inline cond: Boolean, inline msg: String): Unit =
       if xCheckMacro && !cond then
         xCheckMacroAssertFail(msg)
+
+    private inline def xCheckMacroWarning(inline cond: Boolean, inline msg: String): Unit =
+      if xCheckMacro && !cond then
+        val error = new AssertionError()
+        if !yDebugMacro then
+          // start stack trace at the place where the user called the reflection method
+          error.setStackTrace(
+            error.getStackTrace
+              .dropWhile(_.getClassName().startsWith("scala.quoted.runtime.impl")))
+        val stackTrace = error.getStackTrace().map(_.toString).mkString("\n")
+        report.warning(s"$msg\n${stackTrace}")
 
     private def xCheckMacroAssertFail(msg: String): Unit =
       val error = new AssertionError(msg)

--- a/tests/pos-macros/i25541/Macro_1.scala
+++ b/tests/pos-macros/i25541/Macro_1.scala
@@ -1,0 +1,24 @@
+import scala.deriving.Mirror
+import scala.quoted.*
+
+object LocalMacros:
+  def macroRWImpl[T: Type](using Quotes): Expr[Unit] =
+    import quotes.reflect.*
+
+    def inspect(tpe: TypeRepr, typeArgs: List[TypeRepr]): Unit =
+      val constructorSym = tpe.typeSymbol.primaryConstructor
+      val constructorParamSymss = constructorSym.paramSymss
+      val (tparams0, params0) = constructorParamSymss.flatten.partition(_.isType)
+      val constructorTpe = tpe.memberType(constructorSym).widen
+      params0.foreach { param =>
+        constructorTpe.memberType(param).substituteTypes(tparams0, typeArgs)
+      }
+
+    TypeRepr.of[T] match
+      case applied: AppliedType => inspect(TypeRepr.of[T], applied.args)
+      case _ => inspect(TypeRepr.of[T], Nil)
+
+    '{ () }
+
+trait Custom:
+  inline def macroRW[T](using Mirror.Of[T]): Unit = ${ LocalMacros.macroRWImpl[T] }

--- a/tests/pos-macros/i25541/Test_2.scala
+++ b/tests/pos-macros/i25541/Test_2.scala
@@ -1,0 +1,5 @@
+final case class MessageError[Node](message: Option[String], code: Option[Int], data: Option[Node])
+
+object Test:
+  def readWriter(custom: Custom): Unit =
+    custom.macroRW[MessageError[Int]]

